src/main/java/uk/gov/justice/digital/delius/transformers/AppointmentPatchRequestTransformer.java

```diff
@@ -0,0 +1,100 @@
+import uk.gov.justice.digital.delius.config.DeliusIntegrationContextConfig.IntegrationContext;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.StreamSupport;
+
+import static com.fasterxml.jackson.databind.node.TextNode.valueOf;
+import static com.github.fge.jackson.jsonpointer.JsonPointer.of;
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+
+@Component
+@AllArgsConstructor
+public class AppointmentPatchRequestTransformer {
+
+    private static final String JSON_PATCH_PATH_FIELD = "path";
+    private static final String JSON_PATCH_VALUE_FIELD = "value";
+
+    private static final String SOURCE_ATTENDED_FIELD_PATH = "/attended";
+    private static final String SOURCE_NOTIFY_PP_FIELD_PATH = "/notifyPPOfAttendanceBehaviour";
+    private static final String TARGET_OUTCOME_FIELD_NAME = "outcome";
+
+    private final ObjectMapper objectMapper;
+
+    public JsonPatch mapAttendanceFieldsToOutcomeOf(final JsonPatch jsonPatch, final IntegrationContext context) {
+
+        var nodes = objectMapper.convertValue(jsonPatch, JsonNode.class);
+
+        var attendedNode = getAsText(SOURCE_ATTENDED_FIELD_PATH, nodes);
+        var notifyBehaviourNode = getAsBoolean(SOURCE_NOTIFY_PP_FIELD_PATH, nodes);
+
+        var otherReplaceOperations = StreamSupport
+            .stream(nodes.spliterator(), false)
+            .filter(node -> {
+                var fieldPath = node.path(JSON_PATCH_PATH_FIELD).asText();
+                return !SOURCE_ATTENDED_FIELD_PATH.equals(fieldPath) &&
+                    !SOURCE_NOTIFY_PP_FIELD_PATH.equals(fieldPath);
+            })
+            .map(node -> objectMapper.convertValue(node, JsonPatchOperation.class))
+            .collect(toList());
+
+        var replaceOperationsWithOutcome = addReplaceOperationForOutcomeIfAttended(
+            context, attendedNode, notifyBehaviourNode, otherReplaceOperations);
+
+        return new JsonPatch(replaceOperationsWithOutcome);
+    }
+
+    private List<JsonPatchOperation> addReplaceOperationForOutcomeIfAttended(final IntegrationContext context,
+                                                                             final Optional<String> attendedNode,
+                                                                             final Optional<Boolean> notifyBehaviourNode,
+                                                                             final List<JsonPatchOperation> originalReplaceOperations) {
+
+        var replaceOperations = new ArrayList<>(originalReplaceOperations);
+
+        if ( attendedNode.isPresent() ) {
+            var mappings = context.getContactMapping()
+                .getAttendanceAndBehaviourNotifiedMappingToOutcomeType();
+
+            var attendedType = attendedNode.orElseThrow(
+                () -> new IllegalStateException("Attended does not exist in json patch"));
```

Comment:
I don't really understand the reasoning behind choosing `PATCH` so forgive me if my viewpoint is naive... but don't we lose any possibility of documenting business rules like this in the OpenApi documentation by choosing PATCH?

My gut feeling is that it would be better to use a `PUT` or `POST` with a request object that contains the information the client knows about, then we can put the business rules in the API and lose quite a lot of complicated code.  I guess you can work out what should go in the `JsonPatch` by reading the tests but... :man_shrugging:

Reply:
Would your concerns be less if there was a GET endpoint the retrieve the appointment resource? This would give the resource representation that is being patched and possibly make things a lot clearer. The fact that we don't have this means we're in a slightly strange state with the resource modelling of appointments. I take the point that the transformation code is not optimal and we will need to work to improve this.

Reply:
Moved away from patch as made the endpoint a far more business specific operation on an appointment. Agree, much more self documenting and simpler.